File: app/node/audio/volume/volume.cpp
/***

  Olive - Non-Linear Video Editor
  Copyright (C) 2022 Olive Team
  Modifications Copyright (C) 2025 mikesolar

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "volume.h"

#include "widget/slider/floatslider.h"

namespace olive
{

const QString VolumeNode::k_samples_input = QStringLiteral("samples_in");
const QString VolumeNode::k_volume_input = QStringLiteral("volume_in");

#define super MathNodeBase

VolumeNode::VolumeNode()
{
	add_input(k_samples_input, NodeValue::k_samples,
			 InputFlags(k_input_flag_not_keyframable));

	add_input(k_volume_input, NodeValue::k_float, 1.0);
	set_input_property(k_volume_input, QStringLiteral("min"), 0.0);
	set_input_property(k_volume_input, QStringLiteral("view"),
					 FloatSlider::k_decibel);

	set_flag(k_audio_effect);
	set_effect_input(k_samples_input);
}

QString VolumeNode::name() const
{
	return tr("Volume");
}

QString VolumeNode::id() const
{
	return QStringLiteral("org.olivevideoeditor.Olive.volume");
}

QVector<Node::CategoryID> VolumeNode::category() const
{
	return { k_category_filter };
}

QString VolumeNode::description() const
{
	return tr("Adjusts the volume of an audio source.");
}

void VolumeNode::value(const NodeValueRow &value, const NodeGlobals &globals,
					   NodeValueTable *table) const
{
	// Create a sample job
	SampleBuffer buffer = value[k_samples_input].to_samples();

	if (buffer.is_allocated()) {
		// If the input is static, we can just do it now which will be faster
		if (is_input_static(k_volume_input)) {
			auto volume = value[k_volume_input].to_double();

			if (!qFuzzyCompare(volume, 1.0)) {
				buffer.transform_volume(volume);
			}

			table->push(NodeValue::k_samples, QVariant::fromValue(buffer), this);
		} else {
			// Requires job
			SampleJob job(globals.time(), k_samples_input, value);
			job.insert(k_volume_input, value);
			table->push(NodeValue::k_samples, QVariant::fromValue(job), this);
		}
	}
}

void VolumeNode::process_samples(const NodeValueRow &values,
								const SampleBuffer &input, SampleBuffer &output,
								int index) const
{
	return process_samples_internal(values, k_op_multiply, k_samples_input,
								  k_volume_input, input, output, index);
}

void VolumeNode::retranslate()
{
	super::retranslate();

	set_input_name(k_samples_input, tr("Samples"));
	set_input_name(k_volume_input, tr("Volume"));
}

}

File: src/node/rust/src/nodes/volume.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! Volume audio effect (C++ `src/node/src/audio/volume/volume.{h,cpp}`,
//! `olive::VolumeNode`).

use crate::factory::NodeMeta;
use crate::node::{Category, NodeBehavior, NodeCore};

/// Samples input id (C++ `k_samples_input`). Type: samples; flags:
/// not-keyframable; this is the node's effect input
/// (`set_effect_input(k_samples_input)`).
pub const SAMPLES_INPUT: &str = "samples_in";

/// Volume input id (C++ `k_volume_input`). Type: float; default `1.0`;
/// properties: `min = 0.0`, `view = decibel`.
pub const VOLUME_INPUT: &str = "volume_in";

/// Volume effect node. Multiplies every channel of a sample buffer by a
/// gain factor.
///
/// The C++ class derives from `MathNodeBase` (for the shared
/// `process_samples_internal` helper used with `k_op_multiply`) but
/// declares no own data members, so this is a unit-like struct; the
/// multiply-by-scalar sample math will be shared with
/// `super::mathbase` when the base module lands.
pub struct VolumeNode;

impl NodeBehavior for VolumeNode {
	/// Human-readable name (C++ `name()`).
	fn name(&self) -> &str {
		"Volume"
	}

	/// Stable type id (C++ `id()`).
	fn type_id(&self) -> &str {
		"org.olivevideoeditor.Olive.volume"
	}

	/// Categories (C++ `category()`).
	fn categories(&self) -> &[Category] {
		&[Category::Filter]
	}

	/// Description (C++ `description()`).
	fn description(&self) -> &str {
		"Adjusts the volume of an audio source."
	}

	/// Localized input names (C++ `retranslate()`): `samples_in` ->
	/// "Samples", `volume_in` -> "Volume".
	fn input_name<'a>(&self, id: &'a str) -> &'a str {
		todo!()
	}

	/// Evaluate outputs (C++ `value()`): unallocated samples input ->
	/// push nothing; allocated samples with a static volume input ->
	/// apply the gain immediately and push the transformed samples,
	/// skipping the transform when the gain is effectively 1.0 (C++
	/// preserves the `!qFuzzyCompare(volume, 1.0)` double semantics:
	/// `abs(volume - 1.0) * 1e12 > min(abs(volume), 1.0)`); allocated
	/// samples with a non-static volume input -> build a `SampleJob`
	/// over `samples_in` + `volume_in` and push it as a samples value.
	fn value(
		&self,
		core: &NodeCore,
		inputs: &crate::value::NodeValueRow,
		time: oakcore_rs::Rational,
		table: &mut crate::value::NodeValueTable,
	) {
		todo!()
	}

	/// Process a span of samples (C++ `process_samples()`): delegates to
	/// `MathNodeBase::process_samples_internal` with `k_op_multiply`,
	/// i.e. each output sample is the input sample multiplied by the
	/// `volume_in` value. The C++ signature receives the input buffer
	/// and a sample index; the Rust trait instead hands over a time
	/// `range` and the destination buffer.
	fn process_samples(
		&self,
		core: &NodeCore,
		inputs: &crate::value::NodeValueRow,
		range: oakcore_rs::TimeRange,
		output: &mut crate::value::SampleBuffer,
	) {
		todo!()
	}

	/// Deep copy (C++ `copy()`).
	fn duplicate(&self, core: &NodeCore) -> Option<Box<dyn NodeBehavior>> {
		todo!()
	}
}

/// Constructor (C++ `VolumeNode::VolumeNode()`): adds `samples_in`
/// (samples, not-keyframable) and `volume_in` (float, default 1.0,
/// min/view properties documented on the constant), sets the
/// audio-effect flag and makes `samples_in` the effect input.
pub fn create() -> (NodeCore, Box<dyn NodeBehavior>) {
	todo!()
}

/// Register this node type (C++ `k_audio_volume` in
/// `factory.cpp::create_from_factory_index`).
pub fn register(meta: &mut Vec<NodeMeta>) {
	meta.push(NodeMeta {
		type_id: "org.olivevideoeditor.Olive.volume",
		name: "Volume",
		categories: &[Category::Filter],
		create,
	});
}
